blockdag/validate.go

```diff
@@ -735,6 +733,33 @@ func (dag *BlockDAG) checkBlockContext(block *util.Block, parents blockSet, flag
 	return nil
 }
 
+func (dag *BlockDAG) checkBlockParents(block *util.Block) error {
+	_, err := lookupParentNodes(block, dag)
+	if err != nil {
+		var ruleErr RuleError
+		if ok := errors.As(err, &ruleErr); ok && ruleErr.ErrorCode == ErrInvalidAncestorBlock {
+			err := dag.addNodeToIndexWithInvalidAncestor(block)
+			if err != nil {
+				return err
+			}
+		}
+		return err
+	}
+	return nil
+}
+
+func (dag *BlockDAG) checkBlockTxsFinalized(block *util.Block, node *blockNode, flags BehaviorFlags) error {
+	fastAdd := flags&BFFastAdd == BFFastAdd || dag.index.NodeStatus(node).KnownValid()
+	if fastAdd {
+		return nil
+	}
+	parents, err := lookupParentNodes(block, dag)
```

Comment:
Why not node.parents? Or actually, you do parents.bluest() a couple line below - instead we should probably use node.selectedParent (Please check me on this one)